apps/app/src/components/LinkWithQuery/LinkWithQuery.tsx

```diff
@@ -0,0 +1,51 @@
+"use client";
+
+import Link, { LinkProps } from "next/link";
+import { ComponentProps } from "react";
+import { useDevFAQRouter } from "../../hooks/useDevFAQRouter";
+
+type Url = LinkProps["href"];
+
+export const createQueryHref = (href: Url, query: Record<string, string>): Url => {
+	if (typeof href === "string") {
+		const { pathname, searchParams, hash } = new URL(href, "http://app.devfaq.localhost:3000");
+		const params = Object.fromEntries(searchParams.entries());
+		const newQuery = { ...params, ...query };
+
+		return {
+			pathname: href.startsWith("/") ? pathname : pathname.slice(1),
+			query: newQuery,
+			...(hash && { hash }),
+		};
+	}
+
+	const hrefQuery = !href.query
+		? {}
+		: typeof href.query === "string"
+		? Object.fromEntries(new URLSearchParams(href.query).entries())
+		: href.query;
+	const newHrefQuery = { ...hrefQuery, ...query };
+
+	return { ...href, query: newHrefQuery };
```

Comment:
Spora część logiki powtarza się między linijkami 12-19 a 22-29, a przecież robią one to samo.
Ja bym podszedł do tego tak:
1. Preferujemy, że `href` zawsze jest obiektem.
2. Jeśli jest stringiem, to zamieniamy go na obiekt (tak jak teraz robisz `new URL(href…` i dalej wykonujemy to samo.